File: pages/base-page.ts
import { Page } from '@playwright/test';

/**
 * Base class for all Page Objects.
 * Subclasses set a `path` and implement `expectLoaded()`.
 */
export abstract class BasePage {
  protected readonly page: Page;
  protected readonly path: string = '/';

  constructor(page: Page) {
    this.page = page;
  }

  /** Navigate to this page using its `path` (relative to baseURL). */
  async navigate(): Promise<this> {
    await this.page.goto(this.path);
    return this;
  }

  /**
   * Each page must implement a load assertion that confirms it is ready
   * for interaction (a visible heading, a stable URL, a known testId, etc.).
   */
  abstract expectLoaded(): Promise<void>;
}

File: experimental/python-prototype/pages/base_page.py
"""
Clase base para todos los Page Objects.

Un Page Object es una clase que representa una página o componente de la
aplicación. Encapsula los locators y las acciones, para que los tests
queden limpios y los cambios de UI se arreglen en un solo lugar.
"""

from playwright.sync_api import Page, expect


class BasePage:
    """Funcionalidad compartida por todas las páginas."""

    # Subclases deben sobreescribir esto con la ruta relativa (ej: "/login")
    path: str = "/"

    def __init__(self, page: Page):
        self.page = page

    def navigate(self):
        """Navega a la página usando self.path + base_url del contexto."""
        self.page.goto(self.path)
        return self

    def expect_loaded(self):
        """
        Verifica que la página cargó correctamente.
        Las subclases deben implementar su criterio (un heading visible,
        un texto, una URL, etc.).
        """
        raise NotImplementedError(
            "Cada Page Object debe implementar expect_loaded()"
        )
